feat: special purpose pages content collection

jurgenbelien opened the pull request:
# Changes

- Lets ESLint allow rest operator for removing properties from objects.
- Another update to error reporting
- Pass noIndex to Pages Content Collection entries
- Add purpose field to Pages Content Collection entries
- Migrates HomePage and NotFoundPage singletons to regular Pages
  -  These pages are set in Website (`app`) settings
  - Two separate migrations, one for schema, one for content make this easier
- Download said settings to `site.json`
- Creates a SpecialPages Contentent Collection for 404, home, and others pages to come
- Drops the Home route, now handled by `pages/[locale][...path]/index.astro`

# Associated issue

Resolves #291

# How to test

<!-- example:
1. Open preview link
2. Navigate to ...
3. Tap on ...
4. Verify that ...
5. Etc ...
-->

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have made sure that my PR is easy to review (not too big, includes comments)
- [x] I have made updated relevant documentation files (in project README, docs/, etc)
- [ ] I have added a decision log entry if the change affects the architecture or changes a significant technology
- [x] I have notified a reviewer

<!-- Please strike through and check off all items that do not apply (rather than removing them) -->

jurgenbelien committed:
feat(datocms): better error reporting by showing query variables

jurgenbelien committed:
feat(content): use any page for homepage

jurgenbelien committed:
feat(content): add noIndex field to page meta

jurgenbelien committed:
style(content): write loadEntry and loadCollection as regular functions

jurgenbelien committed:
feat(pages): use SpecialPages collection to render 404

jurgenbelien committed:
chore(eslint): allow using rest for removing properties from objects

jurgenbelien committed:
feat(datocms): migrate special page singletons to regular pages

jurgenbelien committed:
feat: remove Home singleton, hardcode home page

jurgenbelien committed:
feat(scripts): download app settings in prep

jurgenbelien commented on config/datocms/migrations/1752588127_featSpecialPages.ts:
This is the schema migration

jurgenbelien commented on config/datocms/migrations/1752588128_featSpecialPagesContent.ts:
This is the content migration.

MarleenEliza reviewed (commented):
Looks good!

But I have some concerns:
1. The slug of a page cannot be left empty. Isn't this desirable for the Home Page though? The language buttons in the top right corner still reflect the link with the set slug.
<img width="1364" height="563" alt="Screenshot 2025-07-21 at 11 28 58" src="https://github.com/user-attachments/assets/9be8c071-1a6d-40e3-b6c8-f3aa93e38323" />
<img width="1363" height="493" alt="Screenshot 2025-07-21 at 13 21 01" src="https://github.com/user-attachments/assets/8a3fb0d5-267e-479e-a75f-cecd7291bda9" />

2. If the specialized pages are unset, the website cannot be built. I know this is by design, but this might be worth mentioning in the docs?

jurgenbelien committed:
test(content): add test for Pages collection

jurgenbelien committed:
test(content): verify noIndex and purpose behavior

jurgenbelien committed:
docs: remove references to Home Page model

jurgenbelien committed:
docs: explain how to add required Home Page and Not Found Page settings

jurgenbelien commented:
> Looks good!
> 
> But I have some concerns:
> 
> 1. The slug of a page cannot be left empty. Isn't this desirable for the Home Page though? The language buttons in the top right corner still reflect the link with the set slug.
I resolved the issue with the language toggle, and added a test scenario for it. The slug field now has an help text explaining that for home it is ignored. 
> 2. If the specialized pages are unset, the website cannot be built. I know this is by design, but this might be worth mentioning in the docs?
I've added it to `getting-started.md`

jurgenbelien commented on config/datocms/migrations/1752588128_featSpecialPagesContent.ts:
```diff
@@ -0,0 +1,196 @@
+    { nested: true },
+  );
+
+  for (const record of recordsReferencingHomePage) {
+    const { id, type, item_type, meta, creator, ...content } = record;
+    const title = isLocalizedField(content.title)
+      ? content.title[locales[0]]
+      : content.title;
+    console.log(`Updating record "${title || id}"`);
+    await client.items.update(
+      record.id,
+      replaceReferences(content, homePageModel.singleton_item.id, home_page),
+    );
+  }
```
this migrates any record that has a reference to the old home page to the new home page

jurgenbelien commented on config/datocms/migrations/1752588128_featSpecialPagesContent.ts:
```diff
@@ -0,0 +1,196 @@
+    { nested: true },
+  );
+
+  for (const record of recordsReferencingNotFoundPage) {
+    const { id, type, item_type, meta, creator, ...content } = record;
+    const title = isLocalizedField(content.title)
+      ? content.title[locales[0]]
+      : content.title;
+    console.log(`Updating record "${title || id}"`);
+    await client.items.update(
+      record.id,
+      replaceReferences(content, notFoundPageModel.singleton_item.id, not_found_page),
+    );
+  }
```
same for the not found pages, although the chance of that being necessary is slim

bazottie commented on config/datocms/migrations/1752588127_featSpecialPages.ts:
```diff
@@ -0,0 +1,112 @@
+  console.log(
+    'Update Slug field "Slug" (`slug`) in model "\uD83D\uDCD1 Page" (`page`)',
+  );
+  await client.fields.update('SFfpl7xiQaGe5-jy6tlPfg', {
```
```suggestion
  await client.fields.update('page::slug', {

```

jbmoelker commented:
Stale and outdated. Will remain on the roadmap. Closing for now.
